src/encrypt.rs

```diff
@@ -0,0 +1,464 @@
+            COSEAlgorithm::AesCcm64_128_256 => 16,
+        }
+    }
+
+    fn openssl_cipher(value: i8) -> Option<Cipher> {
+        Some(match value {
+            1 => Cipher::aes_128_gcm(),
+            2 => Cipher::aes_192_gcm(),
+            3 => Cipher::aes_256_gcm(),
+            30 => Cipher::aes_128_ccm(),
+            31 => Cipher::aes_256_ccm(),
+            _ => return None,
+        })
+    }
+
+    fn from_nid(nid: openssl::nid::Nid) -> Option<COSEAlgorithm> {
+        Some(match nid {
+            openssl::nid::Nid::AES_128_GCM => COSEAlgorithm::A128GCM,
+            openssl::nid::Nid::AES_192_GCM => COSEAlgorithm::A192GCM,
+            openssl::nid::Nid::AES_256_GCM => COSEAlgorithm::A256GCM,
+            openssl::nid::Nid::AES_128_CCM => COSEAlgorithm::AesCcm16_128_128,
+            openssl::nid::Nid::AES_256_CCM => COSEAlgorithm::AesCcm16_128_256,
+            _ => return None,
+        })
+    }
+}
+
+///  Implementation of the Enc_structure structure as defined in
+///  [RFC8152](https://tools.ietf.org/html/rfc8152#section-5.3).
+///
+///  The encryption algorithm for AEAD algorithms is fairly simple.  The
+///  first step is to create a consistent byte stream for the
+///  authenticated data structure.  For this purpose, we use an
+///  Enc_structure.  The Enc_structure is a CBOR array.  The fields of the
+///  Enc_structure in order are:
+///
+///  1.  A text string identifying the context of the authenticated data
+///      structure.  The context string is:
+///
+///         "Encrypt0" for the content encryption of a COSE_Encrypt0 data
+///         structure.
+///
+///         "Encrypt" for the first layer of a COSE_Encrypt data structure
+///         (i.e., for content encryption).
+///
+///         "Enc_Recipient" for a recipient encoding to be placed in an
+///         COSE_Encrypt data structure.
+///
+///         "Mac_Recipient" for a recipient encoding to be placed in a
+///         MACed message structure.
+///
+///         "Rec_Recipient" for a recipient encoding to be placed in a
+///         recipient structure.
+///
+///  2.  The protected attributes from the body structure encoded in a
+///      bstr type.  If there are no protected attributes, a bstr of
+///      length zero is used.
+///
+///  3.  The protected attributes from the application encoded in a bstr
+///      type.  If this field is not supplied, it defaults to a zero-
+///      length bstr.  (See Section 4.3 for application guidance on
+///      constructing this field.)
+///
+///  The CDDL fragment that describes the above text is:
+///
+///  Enc_structure = [
+///      context : "Encrypt" / "Encrypt0" / "Enc_Recipient" /
+///          "Mac_Recipient" / "Rec_Recipient",
+///      protected : empty_or_serialized_map,
+///      external_aad : bstr
+///  ]
+#[derive(Debug, Clone, Deserialize)]
+struct EncStructure {
+    /// context: "Encrypt0" / "Encrypt" / "Enc_Recipient" / "Mac_Recipient" / "Rec_Recipient"
+    context: String,
```

Comment:
Context could also be an enum that Serializes/Deserialized to String, but there's no large benefit to adding it.